LDMP/models/algorithms.py

```diff
@@ -0,0 +1,138 @@
+"""
+/***************************************************************************
+ LDMP - A QGIS plugin
+ This plugin supports monitoring and reporting of land degradation to the UNCCD 
+ and in support of the SDG Land Degradation Neutrality (LDN) target.
+                              -------------------
+        begin                : 2021-02-25
+        git sha              : $Format:%H$
+        copyright            : (C) 2017 by Conservation International
+        email                : trends.earth@conservation.org
+ ***************************************************************************/
+"""
+
+__author__ = 'Luigi Pirelli / Kartoza'
+__date__ = '2021-03-03'
+
+from typing import Optional, Union, List
+from enum import Enum
+import abc
+
+from marshmallow import Schema, fields, post_load
+
+class AlgorithmNodeType(Enum):
+    Group = 1,
+    Leaf = 2
+
+
+class AlgorithmRunMode(Enum):
+    NotApplicable = 0,
+    Locally = 1,
+    Remotely = 2,
+    Both = 3
+
+
+class AlgorithmBase(abc.ABC):
+
+    def __init__(self,
+            name: str,
+            name_details: Optional[str] = None,
+            group: 'AlgorithmGroup' = None, # string because forwarde declared class
+        ) -> None:
+        super().__init__()
+        self.type: Optional[AlgorithmNodeType] = None
+        self.group = group
+        self.name = name
+        self.name_details = name_details
+
+    def getGroup(self) -> 'AlgorithmBase':
+        return self.group
+
+    def row(self) -> int:
+        if not self.group:
+            # e.g. root node/group
+            return 0
+        
+        # look for self in parent group.algorithms
+        try:
+            return self.group.algorithms.index(self)
+        except ValueError as ex:
+            # something strange happen. Can't find myself
+            return 0
+
+    @abc.abstractmethod
+    def rowCount(self) -> int:
+        pass
+
+    @abc.abstractmethod
+    def columnCount(self) -> int:
+        pass
+
+    @abc.abstractmethod
+    def child(self, row) -> Union['AlgorithmBase', None]: # string because AlgorithmBase not yet completly defined
```

Comment:
A small nitpick, but I guess `Optional` is clearer than `Union` here

```suggestion
    def child(self, row) -> Optional['AlgorithmBase']:  # string because AlgorithmBase not yet completly defined
```